cdap-ranger/cdap-ranger-binding/src/test/java/co/cask/cdap/security/authorization/ranger/binding/RangerAuthorizerTest.java

```diff
@@ -0,0 +1,154 @@
+                    Action.EXECUTE);
+
+    // Test dataset type and module specifically as they can contain . in their names.
+    authorizer.enforce(NAMESPACE_DEFAULT.datasetType("co.cask.table_1_mod"), DEREK, Action.READ);
+    testEnforceFail(new NamespaceId("anotherns").datasetType("co.cask.table_1_mod"), DEREK, Action.READ);
+    testEnforceFail(NAMESPACE_DEFAULT.datasetType("co.cask"), DEREK, Action.READ);
+
+    authorizer.enforce(NAMESPACE_DEFAULT.datasetModule("co.cask.table_1_mod"), ALBERT, Action.READ);
+    authorizer.enforce(NAMESPACE_DEFAULT.datasetModule("co.cask.table_2_mod"), ALBERT, Action.READ);
+    testEnforceFail(new NamespaceId("anotherns").datasetType("co.cask.table_1_mod"), ALBERT, Action.READ);
+    testEnforceFail(NAMESPACE_DEFAULT.datasetType("co.cask"), ALBERT, Action.READ);
+
+    // Test artifacts: we don't enforce on version for artifacts
+    authorizer.enforce(NAMESPACE_DEFAULT.artifact("DataPipeline", "1.0-SNAPSHOT"), SREE, Action.READ);
+    testEnforceFail(new NamespaceId("anotherns").artifact("DataPipeline", "1.0-SNAPSHOT"), SREE, Action.READ);
+    // cdap entities are case sensitive
+    testEnforceFail(NAMESPACE_DEFAULT.artifact("datapipeline", "1.0-SNAPSHOT"), SREE, Action.READ);
+  }
+
+  @Test
+  public void testWildcardPrivileges() throws Exception {
+    // Test privileges for sagar who has ADMIN on stream:default.*
+    authorizer.enforce(NAMESPACE_DEFAULT.stream("teststream"), SAGAR, Action.ADMIN);
+    authorizer.enforce(NAMESPACE_DEFAULT.stream("anotherstream"), SAGAR, Action.ADMIN);
+    testEnforceFail(NAMESPACE_DEFAULT.stream("teststream"), SAGAR, Action.READ);
+    testEnforceFail(NAMESPACE_DEFAULT, SAGAR, Action.ADMIN);
+    authorizer.isVisible(ImmutableSet.<EntityId>of(NAMESPACE_DEFAULT), SAGAR);
+    testVisibleFail(ImmutableSet.<EntityId>of(new NamespaceId("anothernamespace")), SAGAR);
+    testEnforceFail(new NamespaceId("anothernamespace").stream("teststream"), SAGAR, Action.ADMIN);
+
+    // Test privileges for shankar who has ADMIN on program:default.testapp.workflow.*
+    authorizer.enforce(NAMESPACE_DEFAULT.app("testapp").program(ProgramType.WORKFLOW, "prog1"), SHANKAR, Action.ADMIN);
+    authorizer.enforce(NAMESPACE_DEFAULT.app("testapp").program(ProgramType.WORKFLOW, "prog2"), SHANKAR, Action.ADMIN);
+    testEnforceFail(NAMESPACE_DEFAULT.app("testapp").program(ProgramType.MAPREDUCE, "prog1"), SHANKAR, Action.ADMIN);
+    testEnforceFail(NAMESPACE_DEFAULT.app("testapp").program(ProgramType.WORKFLOW, "prog1"), SHANKAR, Action.READ);
+    testEnforceFail(NAMESPACE_DEFAULT.app("testapp"), SHANKAR, Action.ADMIN);
+    authorizer.isVisible(ImmutableSet.<EntityId>of(NAMESPACE_DEFAULT.app("testapp")), SHANKAR);
+    authorizer.isVisible(ImmutableSet.<EntityId>of(NAMESPACE_DEFAULT), SHANKAR);
+    testVisibleFail(ImmutableSet.<EntityId>of(new NamespaceId("anothernamespace")), SHANKAR);
```

Comment:
I think it is better to add a test case for the wildcard of program:ns1.app1.*

Reply:
Done.